ENH replace loss module Gradient boosting

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Closes #25964.

#### What does this implement/fix? Explain your changes.
This replaces the losses from `_gb_losses.py` with the once of the common private loss submodule `sklearn._loss` in a very backward compatible way. Some factors of the 2 loss implementations differ by a factor of 2. Those factors are here accounted for.

#### Any other comments?
1. No use of parallelism of the new loss functions is activated yet. This is deferred to later PRs (and contributor capacity).
2. Should we discuss the backward compatibility for attributes `oob_improvement_` and `train_score_` wrt the above mentioned constant factor of the loss? As `oob_scores_` and `oob_score_` is about to be introduced with 1.3, we could still change them.

lorentzenchr committed:
ENH common losses in gradient boosting

lorentzenchr committed:
MNT replace _validate_y by _encode_y

* Make it similar to HGBT

lorentzenchr committed:
MNT rename _is_initialized to _is_fitted

* Make it more similar to HGBT

lorentzenchr committed:
ENH use common loss function module in GB

lorentzenchr committed:
MNT fix backwards compatibility for oob_scores_ and oob_improvement_

lorentzenchr committed:
FIX test_non_uniform_weights_toy_edge_case_reg

Note that in the case raw_prediction == y, we now have a different sign in AbsoluteError.gradient

lorentzenchr committed:
FIX _init_raw_predictions detection for predict_proba

lorentzenchr committed:
FIX huber GBT

lorentzenchr committed:
FIX factor of losses for backward compat

lorentzenchr committed:
TST add tests for backward compat

lorentzenchr committed:
FIX factor multinomial loss

lorentzenchr committed:
TST add test_binomial_vs_alternative_formulation

lorentzenchr committed:
MNT remove _gb_losses.py and tests

lorentzenchr commented:
A very simple benchmark for binary classification gives:
```
PR:   768 ms ± 9.76 ms
main: 937 ms ± 128 ms
```


<details>

```python
import numpy as np
from sklearn.ensemble import GradientBoostingClassifier
n_samples, n = 10, 10_000
y = np.tile(np.arange(n_samples) % 2, n)
x1 = np.minimum(y, n_samples / 2)
x2 = np.minimum(-y, -n_samples / 2)
X = np.c_[x1, x2]
%timeit GradientBoostingClassifier(n_estimators=100).fit(X, y)
```

</details>

lorentzenchr committed:
DOC add whatsnew entry

lorentzenchr committed:
Merge branch 'main' into gradient_boosting_common_loss

lorentzenchr committed:
FIX don't use sample_weight in gradient

lorentzenchr committed:
CLN remove comment left overs

lorentzenchr committed:
CLN add todo note on use of loss_gradient

thomasjpfan commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -981,6 +982,44 @@ def test_binomial_and_multinomial_loss(global_random_seed):
     )
 
 
+def test_binomial_vs_alternative_formulation():
```
Given the size of this PR, can this test be moved to it's own PR?

lorentzenchr replied:
Yes, I'll do that. Though, without the context of this PR, it would have been an odd PR.

lorentzenchr replied:
Done in #26298.

thomasjpfan commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -981,6 +982,44 @@ def test_binomial_and_multinomial_loss(global_random_seed):
+    """Tast that both formulations of the binomial deviance agree.
+
+    Often, the binomial deviance or log loss is written in terms of a variable
+    z in {-1, +1}, but we use y in {0, 1}, hence y = 2 * y - 1.
```
Did you mean z here?

```suggestion
    z in {-1, +1}, but we use y in {0, 1}, hence z = 2 * y - 1.
```

lorentzenchr replied:
Yes

thomasjpfan commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -981,6 +982,44 @@ def test_binomial_and_multinomial_loss(global_random_seed):
 
 
+def test_binomial_vs_alternative_formulation():
+    """Tast that both formulations of the binomial deviance agree.
```
```suggestion
    """Test that both formulations of the binomial deviance agree.
```

lorentzenchr committed:
Revert "TST add test_binomial_vs_alternative_formulation"

This reverts commit 64098db933889b6201820c21131ad55418a8afb2.

lorentzenchr committed:
TST reduce tol

lorentzenchr committed:
FIX first check sample_weight then encode y

lorentzenchr commented:
> Given the size of this PR ...

Therefore, I provided the different `test_XXX_exact_backward_compat` tests.

lorentzenchr committed:
TST reduce tol

lorentzenchr committed:
CLN minor code cleanups in _update_terminal_regions

lorentzenchr committed:
DOC cross_val_score in common pitfalls

The exact origin for the necessity of this change is unclear. The train_score_ of the GBT inside the pipeline is exactly the same for this branch and 1.2.2.

lorentzenchr commented:
As info, I do not really understand the necessity of https://github.com/scikit-learn/scikit-learn/pull/26278/commits/c6cefe46066dedead4396a8bcaf168405b13773a. See also the detailed commit message:
> The exact origin for the necessity of this change is unclear. The train_score_ of the GBT inside the pipeline is exactly the same for this branch and 1.2.2.

lorentzenchr committed:
FIX use percentile from fixes

lorentzenchr committed:
TST adapt test_non_uniform_weights_toy_edge_case_reg

lorentzenchr committed:
MNT do not unnecessarily copy raw_predictions

lorentzenchr committed:
MNT remove X_csr and X_csc

lorentzenchr committed:
CLN better comment

lorentzenchr committed:
DOC list n_trees_per_iteration_ under Attributes

lorentzenchr committed:
TST skip_if_32bit test_huber_exact_backward_compat

lorentzenchr committed:
Merge branch 'main' into gradient_boosting_common_loss

lorentzenchr commented:
I don't know if time and review capacity permit it, but let's aim for 1.3.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -23,33 +23,246 @@
+        for multiclass classification. ``raw_predictions`` is casted
+        into float64.
+    """
+    # TODO: Use loss.fit_intercept_only where appropriate instead of
+    # DummyRegressor.
```
I can see that `estimator` is `Dummy*`, but it is hard to tell without knowing how `_init_raw_predictions` is used. Can this comment be expanded with this context?

lorentzenchr replied:
Better with https://github.com/scikit-learn/scikit-learn/pull/26278/commits/0c1c552bac8b695c399fcd270dfb8e904d3b5b8e?

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -23,33 +23,246 @@
+                residual_ = residual.take(indices, axis=0)
+                prob = y_ - residual_
+                # numerator = negative gradient = y - prob
+                numerator = np.average(residual_, weights=sw)
```
`main` uses `np.sum(sw * residual)` for the numerator. What is the reason for using average here?

(This question applies to all other places.)

lorentzenchr replied:
It's equivalent and better underlines the fact that when working with weights, average should be called in most circumstances, e.g. `average` instead of `mean`.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -23,33 +23,246 @@
+        return loss.link.link(predictions)
+
+
+def _update_terminal_regions(
```
The original `_gb_losses.py` had this abstraction in the loss classes themselves. Given that this is very specific to trees, I am okay with having a big function handle it.

lorentzenchr replied:
I'm also not a big fan of this, but it got the things done. Maybe, someone clever can refactor later in a more beautiful way.

thomasjpfan commented on doc/whats_new/v1.3.rst:
```diff
@@ -293,6 +293,11 @@ Changelog
   out-of-bag scores via the `oob_scores_` or `oob_score_` attributes.
   :pr:`24882` by :user:`Ashwin Mathur <awinml>`.
 
+- |Efficiency| :class:`ensemble.GradientBoostingClassifier` is faster,
```
This needs to be moved to 1.4

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -424,23 +638,28 @@ def fit(self, X, y, sample_weight=None, monitor=None):
             stratify = y if is_classifier(self) else None
-            X, X_val, y, y_val, sample_weight, sample_weight_val = train_test_split(
+            (
+                X_train,
```
I do like changing these variables to `X_train`, etc. To make this PR smaller, may you open a PR that does these variable name changes?

lorentzenchr replied:
I did that change because `X` gets overwritten with `X_train` which I consider a bug. You're right that a separate PR might have been better.

lorentzenchr replied:
See #26959.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -1042,6 +1289,12 @@ class GradientBoostingClassifier(ClassifierMixin, BaseGradientBoosting):
+        The number of trees that are built at each iteration. For regressors, this is
+        always 1.
+
+        .. versionadded:: 1.3.0
```
```suggestion
        .. versionadded:: 1.4.0
```

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -1210,20 +1463,59 @@ def __init__(
+                )
+        elif self.loss == "exponential":
+            if self.n_classes_ > 2:
+                raise ValueError(
```
There are some codecov issues with `_get_loss`.

lorentzenchr replied:
https://github.com/scikit-learn/scikit-learn/pull/26278/commits/40315f58b6d795a01bc88285518afa37a39635c5

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -197,8 +410,12 @@ def __init__(
+
+    @abstractmethod
+    def _get_loss(self, sample_weight):
+        pass
```
For codecov to be happy:

```suggestion
        """Get loss object from sklearn._loss.loss."""
```

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -23,33 +23,246 @@
+        # Same as np.quantile(abserr, loss.quantile, axis=0, method="inverted_cdf")
+        delta = percentile(abserr, 100 * loss.quantile, axis=0, method="lower")
+    else:
+        delta = _weighted_percentile(abserr, sample_weight, 100 * loss.quantile)
```
Very strange to not have coverage here. If you have a chance, can you open a new PR to add a test to check huber loss & sample_weights?

lorentzenchr replied:
I opened the issue #26960.

thomasjpfan reviewed (commented):
Here is another pass. I see a lot of familiar patterns from `HistGradientBoosting*`.

lorentzenchr committed:
CLN expand comment

lorentzenchr committed:
Merge branch 'main' into gradient_boosting_common_loss

lorentzenchr committed:
DOC mvoe whatsnew to 1.4

lorentzenchr committed:
CLN increase versionadded to 1.4

lorentzenchr committed:
CLN make codecov happy

lorentzenchr committed:
ENH add BaseLoss to param constraint and add tests

lorentzenchr committed:
Merge branch 'main' into gradient_boosting_common_loss

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -28,19 +28,231 @@
+                denominator = np.average(prob * (1 - prob), weights=sw)
+
+                # prevents overflow and division by zero
+                sw_sum = indices.shape[0] if sw is None else np.sum(sw)
+                if abs(denominator * sw_sum) < 1e-150:
```
I know this is trying to reproduce the original code by multipling by `np.sum(sw)`, but I think we can get away with just catching the warning:

```python
def _safe_divide(numerator, denominator):
    with np.errstate(divide="raise"):
        try:
            return numerator / denominator
        except FloatingPointError as e:
            return 0.0
```

and then using `_safe_divide` everywhere.

(Given that codecov is complaining, we do not even test for this edge case.)

lorentzenchr replied:
Good idea, done in 2f0a2930ca291ddb961464800c372cf9acc804db.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -203,25 +419,30 @@ def _fit_stage(
-        raw_predictions_copy = raw_predictions.copy()
+        if isinstance(self._loss, HuberLoss):
+            set_huber_delta(loss=self._loss, y_true=y, raw_prediction=raw_predictions)
+        # TODO: Without oob, i.e. with self.subsample = 1, we could call
```
To make it clear it is a proportion:

```suggestion
        # TODO: Without oob, i.e. with self.subsample = 1.0, we could call
```

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -653,7 +890,9 @@ def _fit_stages(
                 # By calling next(y_val_pred_iter), we get the predictions
                 # for X_val after the addition of the current stage
-                validation_loss = loss_(y_val, next(y_val_pred_iter), sample_weight_val)
+                validation_loss = self._loss(
```
Does this loss require a factor?

lorentzenchr replied:
Yes, but almost no. The factor in this place does not change the behavior / program execution and the `validation_loss` is not made public anywhere.
Nevertheless, I added it.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -305,9 +512,18 @@ def _init_state(self):
         self.init_ = self.init
         if self.init_ is None:
-            self.init_ = self._loss.init_estimator()
+            if is_classifier(self):
```
Compared to `main`, I think selecting the initial estimator and selecting terminal nodes are worst codewise. Do you think it's worth making subclasses of the loss functions to bring these abstractions back?

```python
class HalfBinomialLossGB(HalfBinomialLoss):
    def init_estimator(...):
        ...
    def update_terminal_regions(...):
        ...

lorentzenchr replied:
I would really like to keep it the way it is.
I hope the code will come closer to `_update_leaves_values` in the HGBT code some day.

lorentzenchr replied:
I tried to make it a bit nicer with https://github.com/scikit-learn/scikit-learn/pull/26278/commits/f1f1ccb16fa31165e8e1644ed8430f27bfcdbf2a, mainly pulling the logic out of the loop over leaves.

thomasjpfan replied:
I agree that [`f1f1ccb` (#26278)](https://github.com/scikit-learn/scikit-learn/pull/26278/commits/f1f1ccb16fa31165e8e1644ed8430f27bfcdbf2a) is a bit nicer. Let's go with that.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -1211,20 +1456,60 @@ def __init__(
+        # expose n_classes_ attribute
+        self.n_classes_ = n_classes
+        if sample_weight is None:
+            n_trim_classes = n_classes
```
`fit` always passes in a non-None `sample_weight` to `_encode_y`. To go down this code path, `fit` needs to pass `sample_weight_is_none` to `_encode_y`.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -203,25 +419,30 @@ def _fit_stage(
-            residual = loss.negative_gradient(
-                y, raw_predictions_copy, k=k, sample_weight=sample_weight
-            )
+            # Note: We need the negative gradient!
+            if gradient.ndim == 2:
+                residual = -gradient[:, k]
+            else:
+                residual = -gradient
```
Can this negation be done outside the loop?

lorentzenchr committed:
ENH _save_divide

lorentzenchr committed:
CLN neg_gradients outside of main loop

lorentzenchr committed:
CLN _encode_y and  validation_loss review comments

lorentzenchr committed:
ENH handle gradient.ndim=2 as in HGBT

lorentzenchr committed:
CLN add TODO where to put in the learning rate

lorentzenchr committed:
CLN correct docstring of _update_terminal_regions

lorentzenchr committed:
CLN rename residual to neg_gradient

lorentzenchr committed:
CLN tree value update

lorentzenchr committed:
ENH add compute_update to pull if-else of losses out of leaf loop

lorentzenchr committed:
TST add test_raise_if_init_has_no_predict_proba

lorentzenchr committed:
Merge branch 'main' into gradient_boosting_common_loss

lorentzenchr committed:
ENH remove unweighted set_huber_delta

lorentzenchr committed:
CLN use check_is_fitted and remove NotFittedError

lorentzenchr commented:
@thomasjpfan It's ready again (codecov is green!), meaning that I don't plan to change more in this PR except for review comments.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -1211,20 +1477,60 @@ def __init__(
+            n_trim_classes = n_classes
+        else:
+            n_trim_classes = np.count_nonzero(
+                np.bincount(encoded_y.astype(int), sample_weight)
```
A few lines above where the `LabelEncoder` is called, can the variable be called `encoded_y_int`:

```python
label_encoder = LabelEncoder()
encoded_y_int = label_encoder.fit_transform(y)
```

This way, we can reuse it here without having to convert `encoded_y` back into an int.

lorentzenchr replied:
good point

lorentzenchr commented on sklearn/ensemble/_gb.py:
```diff
@@ -1160,7 +1426,7 @@ class GradientBoostingClassifier(ClassifierMixin, BaseGradientBoosting):
     _parameter_constraints: dict = {
         **BaseGradientBoosting._parameter_constraints,
-        "loss": [StrOptions({"log_loss", "exponential"})],
+        "loss": [StrOptions({"log_loss", "exponential"}), BaseLoss],
```
Fine.
May I introduce it with another PR?

thomasjpfan replied:
I am see adding `BaseLoss` as a follow up to this PR. 

Although, I am not sure if we should support custom `BaseLoss`. Even for this PR, `_update_terminal_regions` and `self.init_`  is loss dependent and can not be adjusted with just passing in a `BaseLoss`.

Do you have a use case for allowing for `BaseLoss`?

lorentzenchr replied:
Good points. It sometimes helps testing, but not here. In HGBT, it enables to _privately_ pass ones own loss function. I discard the idea of adding it here.

lorentzenchr committed:
CLN remove test_get_loss and do not allow BaseLoss as argument

lorentzenchr committed:
CLN use encoded_y_int

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -418,34 +642,40 @@ def fit(self, X, y, sample_weight=None, monitor=None):
+        self._set_max_features()
 
-        self._check_params()
+        if isinstance(self.loss, str):
+            self._loss = self._get_loss(sample_weight=sample_weight)
+        elif isinstance(self.loss, BaseLoss):
+            self._loss = self.loss
```
Without `BaseLoss` support, can we safely do this:

```suggestion
        self._loss = self._get_loss(sample_weight=sample_weight)
```

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -1211,20 +1477,58 @@ def __init__(
+            else:
+                return ExponentialLoss(sample_weight=sample_weight)
+        else:
+            raise ValueError(
```
With the parameter validation, this `else` block never runs, thus codecov is complaining.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -203,27 +418,39 @@ def _fit_stage(
-        assert sample_mask.dtype == bool
-        loss = self._loss
+        """Fit another stage of ``n_trees_per_iteration_`` trees."""
+        if sample_mask.dtype != bool:
+            raise ValueError("Only dtype bool is allowed for sample_mask.")
```
This is never called because `_fit_stages` always constructions `sample_mask` properly, thus codecov is complaining.

I am okay with leaving the `assert` or removing the check all together.

lorentzenchr replied:
I'll remove it.

thomasjpfan commented on sklearn/ensemble/_gb.py:
```diff
@@ -28,19 +28,230 @@
+        try:
+            return numerator / denominator
+        except FloatingPointError:
+            return 0.0
```
For codecov, can we add a small test to trigger the divide by zero?

lorentzenchr replied:
I tried to find a case with with GB estimator, but that's hard. So I simply added `test_safe_divide`.

glemaitre replied:
Apparently, we have a nice example that check this one ;)

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
TST test_safe_divide

lorentzenchr committed:
Merge branch 'main' into gradient_boosting_common_loss

lorentzenchr committed:
MNT remove test_probability_exponential

- This test is not necessary anymore.
  Already tested in the _loss module.

thomasjpfan reviewed (approved):
Thank you for the updates! LGTM

lorentzenchr commented:
Dear future 2nd reviewer
This is much easier to review than it looks by just having a look at the tests.

OmarManzoor reviewed (commented):
Thanks for the PR @lorentzenchr. A few comments.

lorentzenchr committed:
ENH improve compute_update for ExponentialLoss

lorentzenchr commented on sklearn/ensemble/_gb.py:
```diff
@@ -28,19 +28,229 @@
+            def compute_update(y_, indices, neg_gradient, raw_prediction, k):
+                z = 2.0 * y_ - 1.0  # z is -1 or +1
+                raw_pred = raw_prediction[indices, k]
+                # numerator = negative gradient
+                numerator = np.average(z * np.exp(-z * raw_pred), weights=sw)
```
Good point.

lorentzenchr commented on sklearn/ensemble/_gb.py:
```diff
@@ -28,19 +28,229 @@
+            # TODO: Multiply here by learning rate instead of everywhere else.
+            tree.value[leaf, 0, 0] = update
+
+    # update predictions (both in-bag and out-of-bag)
```
They are updated but they don't change their value. Therefore, they are not updated:smirk:
I'll add a comment.

lorentzenchr committed:
DOC SquaredError specialties in update terminal regions

lorentzenchr committed:
CLN address review comments

lorentzenchr commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1120,39 +1139,23 @@ def test_warm_start_wo_nestimators_change():
+    [
+        ("squared_error", 0.5),
+        ("absolute_error", 0.0),
+        ("huber", 0.0),
```
Well spotted. I changed it back to 0.5.

OmarManzoor commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1120,39 +1139,23 @@ def test_warm_start_wo_nestimators_change():
+    ("loss", "value"),
+    [
+        ("squared_error", 0.5),
+        ("absolute_error", 0.0),
```
"absolute_error" still has a value of 0.0. If this is required for this particular loss, then it seems fine. Otherwise why not just remove "value" from the parametrization and set `gb.predict([[1, 0]])[0] >= 0.5`?

lorentzenchr replied:
That's needed. I tested 0, but it does not work.

lorentzenchr replied:
I guess that was the reason I introduces the parametrization with loss and value.

OmarManzoor replied:
Sounds good. Thank you for clarifying.

OmarManzoor replied:
However we also do not have a backward compatibility test for absolute error. Would it make sense to add that?

lorentzenchr replied:
If you insist, then yes. It'll take some time for me to do that.

OmarManzoor replied:
No I don't insist 😄 . I think there are other tests that aren't changed, which do use `abolute_error`, so hopefully that should prove that the results are consistent.

OmarManzoor replied:
Specifically this one
https://github.com/scikit-learn/scikit-learn/blob/51ca717db5065e912cdd6cd838bb08d4d08c2770/sklearn/ensemble/tests/test_gradient_boosting.py#L120-L158

OmarManzoor commented on sklearn/ensemble/_gb.py:
```diff
@@ -1211,20 +1477,53 @@ def __init__(
+        return encoded_y
+
+    def _get_loss(self, sample_weight):
+        if self.loss == "log_loss":
+            if self.n_classes_ == 2:
+                return HalfBinomialLoss(sample_weight=sample_weight)
+            else:
+                return HalfMultinomialLoss(
+                    sample_weight=sample_weight, n_classes=self.n_classes_
+                )
+        elif self.loss == "exponential":
+            if self.n_classes_ > 2:
+                raise ValueError(
+                    f"loss='{self.loss}' is only suitable for a binary classification "
+                    f"problem, you have n_classes={self.n_classes_}. "
+                    "Please use loss='log_loss' instead."
+                )
+            else:
+                return ExponentialLoss(sample_weight=sample_weight)
```
In all three of these losses sample_weight is not being used inside the init methods. Do we really need to pass it here? If it is done to maintain consistency then I think it is fine.

lorentzenchr replied:
That's true. The point is that I like no-brain solutions and I do not want to care about the internals of the losses. So let's just pass the sample weights here.

OmarManzoor reviewed (commented):
Thank you for the updates @lorentzenchr! Looks good now. Just a few minor comments.

OmarManzoor reviewed (approved):
LGTM. Thanks @lorentzenchr

lesteve commented:
It looks like one of the example broke after merging this PR, namely `examples/ensemble/plot_gradient_boosting_regularization.py`, because of NaNs.

From [build log](https://app.circleci.com/pipelines/github/scikit-learn/scikit-learn/50213/workflows/b3198c4b-b815-4481-9290-e506ab8d44ba/jobs/247550):
```
Unexpected failing examples:
/home/circleci/project/examples/ensemble/plot_gradient_boosting_regularization.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/project/examples/ensemble/plot_gradient_boosting_regularization.py", line 77, in <module>
    test_deviance[i] = 2 * log_loss(y_test, y_proba[:, 1])
  File "/home/circleci/project/sklearn/utils/_param_validation.py", line 211, in wrapper
    return func(*args, **kwargs)
  File "/home/circleci/project/sklearn/metrics/_classification.py", line 2831, in log_loss
    y_pred = check_array(
  File "/home/circleci/project/sklearn/utils/validation.py", line 958, in check_array
    _assert_all_finite(
  File "/home/circleci/project/sklearn/utils/validation.py", line 123, in _assert_all_finite
    _assert_all_finite_element_wise(
  File "/home/circleci/project/sklearn/utils/validation.py", line 172, in _assert_all_finite_element_wise
    raise ValueError(msg_err)
ValueError: Input contains NaN.
```

glemaitre commented:
Actually, the only case that go side ways is when both the numerator and denominator goes to 0.0. I assume that it can happen when at the previous iteration we return 0.0 because of the `FloatingPointError`.
